doma-processor/src/main/java/org/seasar/doma/internal/apt/decl/TypeDeclaration.java

```diff
@@ -226,6 +217,32 @@ private void removeHiddenFieldDeclarations(List<FieldDeclaration> candidates) {
     }
   }
 
+  public List<MethodDeclaration> getScopeMethods(ClassName metamodel) {
+    return getMethods(m -> isScopeMethod(m, metamodel));
+  }
+
+  public boolean isScopeMethod(ExecutableElement m, ClassName metamodel) {
+    if (m.getModifiers().contains(Modifier.STATIC)) {
+      return false;
+    }
+
+    if (!m.getModifiers().contains(Modifier.PUBLIC)) {
+      return false;
+    }
+
+    if (m.getReturnType().getKind() == TypeKind.VOID) {
+      return false;
+    }
+
+    if (m.getParameters().size() < 1) {
+      return false;
+    }
+
+    VariableElement firstParameter = m.getParameters().get(0);
+    // TODO: この時点ではmetamodelが作られていないので代入可能かのチェックができない。SimpleNameの比較をしているが名前空間がちがえばすり抜けるため他の良い方法があれば
+    return firstParameter.asType().toString().equals(metamodel.getSimpleName());
```

Comment:
Here, type checking cannot be performed correctly because it is before the Metamodel is generated.
Do you have a good idea?

Reply:
I have no idea.
But I think it's OK.